kroxylicious-filters/kroxylicious-encryption/doc/design.adoc

```diff
@@ -0,0 +1,267 @@
+<4> Computed from `baseSequence` + initial offset within the batch
++
+The number of octets for each terminal are dictated by the Kafka message format. See <https://kafka.apache.org/documentation/#messageformat>
+
+== Constructing the encrypted record
+
+The "encrypted record" is the record which actually gets appended to the log.
+
+First let's define the _wrapping_:
+
+[source,abnf]
+.v1 wrapping schema
+----
+wrapping-v1                = edek-length
+                             edek
+                             aad-code
+                             cipher-code
+                             cipher-specific-data
+                             parcel-ciphertext-length
+                             parcel-ciphertext
+edek-length                = 1*OCTET      ; unsigned VARINT
+edek                       = *OCTET       ; edek_length bytes
+aad-code                   = OCTET        ; <1>
+cipher-code                = OCTET        ; <2>
+cipher-specific-data       = *OCTET       ; <3>
+iv                         = 12OCTET      ; length implied by v1
+parcel-ciphertext-length   = *OCTET ; VARINT
+parcel-ciphertext          = *OCTET
+----
+<1> 0 means None, 1 means Batch (see AAD section)
+<2> 0 means AES-GCM with 96 bit IV and 128 bit authentication tag
+<3> The schema depends on the cipher. For `cipher-code=1` this is just the 12 byte IV.
+
+For reference, this is the schema of a Kafka record:
+
+[source,abnf]
+.Kafka's record schema
+----
+record                     = length
+                             attributes
+                             timestampDelta
+                             offsetDelta
+                             recordKeyLength
+                             recordKey
+                             recordValueLen
+                             recordValue
+                             *recordHeader
+recordHeader               = headerKeyLength
+                             headerKey
+                             headerValueLength
+                             headerValue
+----
+
+To construct the "ciphertext record":
+
+. Construct the `parcel` from the plaintext record's `value` and optionally the `headers`
+. Encrypt the `parcel`, and construct the `wrapping`
+. Create a copy of the plaintext record, but:
+.. replace the `value` with the `wrapping` (adjusting the `valueLen`)
+.. optionally clear the `headers` (if they were included in the parcel)
+.. Prepend a header with name "kroxylicious.io/encryption" and byte value `0x01`.
+
+
+=== Compression
+
+Kafka supports compression of record batches (not individual records). This means there are a number of places where compression could be used:
+
+==== Client sends a compressed batch
+It's _batches_ that are compressed (by the producer), but _records_ that are encrypted, so we have no choice but to decompress the whole batch if we're to encrypt any of the records within it.
+There is the possibility of (re-)compressing records prior to encryption.
```

Comment:
Do you mean compressing the records (individually) before encryption?   Or do you mean compressing the batch of encrypted records?

Reply:
suggest we merge with a comment/todo and iterate when Tom's back